### DISCUSSION TITLE
Git/GitHub storage should be able to clone the entire repo

### DISCUSSION THREAD

**Discussion by znicholasbrown** [PrefectHQ/prefect#4776] [Ideas]
From @nathan5280 in Slack:

> We have a bunch of core logic in it's own package that we are building into a core base image to run our specific flows in using the ECS agent.
> Our specific flows have a fair amount of logic in a couple of modules.   We see in the code that both Git and GitHub storage classes only wind up with the flow.py file getting loaded into the image at runtime.   Granted they get there by two different.  Git clones the repo, extracts the flow.py and then deletes the repo.  GitHub downloads just the flow.py file.
> It seems like it wouldn't be all that difficult to modify the Git storage class to not delete the repo after it extracts the flow.py file and have the root of the modules on PYTHONPATH in the base image.  Then when the flow is executed it would be able to find the modules and everything would be Prefect.
> While we are all noodling on this - is there anything obvious that keeps us from:
> - Creating a new GitRepo storage class that keeps the repo around
> - Adding it to the core package and building it into the core image
> - Adding where that repo's root will be to PYTHONPATH
> - running with the specific models loaded from Git.

cc: @nathan5280, @MSDuncan82, @benjamineverett

**Comment by zanieb** [comment]
This is not necessarily compatible with distributed execution, e.g. if your function is running on a dask worker it won't have access to these python modules or any of the other files in your downloaded repository. We may extend `Git` storage to leave the repo around which would allow something like this when not using distributed execution, but we have held off on doing so because it may create unrealistic expectations and unexpected errors as people try to run their flows in different deployment patterns. Just placing the modules on the PYTHONPATH will bypass any installation of your module requirements (vs `pip install`) which means that there would need to be some sort of containerization anyway which is where we would typically recommend addressing this. Perhaps there would be a straight-forward way to include a `pip install your_package @ git` pointing to the same commit as the `Git` storage in a `DockerRun`. This is something that still needs a significant amount of design. I'd be interseted to hear some more about how/why the current container-based setup you're using isn't working though.

I'm not sure if you can roll your own storage type, cc @cicdw for context on that. This is something that is tricky for us to accept into Prefect core since we cannot guarantee it will work consistently, but I am happy to take a look at an implementation as long as it is very clear to you that it is likely not going to be accepted soon.

**Reply by nathan5280** [reply]
Ahh, I figured there was more to it than I was thinking.   

I think we can keep moving ahead without any problems by using the Docker storage.  This will get all the containerization and dependencies into one image.

We appreciate the concern about things working the same in the different environments ...

Currently, we don't have any dependencies in the specific modules that aren't already in the core.   But you make a fair point about skipping `pip install` and not getting our dependencies installed.  We could hack this by installing them with the core even if they weren't used by core, but that seems a bit kludgy and probably not well received by a future teammate.

In the different environments, which of these will work?
- LocalExecutor - Seems like this will work if there are no additional dependencies and the module is on PYTHONPATH.
- DaskExecutor 
    - Ephemeral in the container.   It seems like this will work the same as the LocalExecutor.
    - Persistent cluster running outside the workflow process.   This won't work unless the specific modules were install on the machines running the cluster.
